src/main/java/org/spongepowered/common/mixin/core/block/material/MixinMapColor.java

```diff
@@ -0,0 +1,128 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.mixin.core.block.material;
+
+import com.google.common.base.Objects;
+import org.spongepowered.api.data.DataContainer;
+import org.spongepowered.api.data.DataQuery;
+import org.spongepowered.api.data.MemoryDataContainer;
+import org.spongepowered.api.data.Queries;
+import org.spongepowered.api.map.color.MapColor;
+import org.spongepowered.api.map.color.MapShade;
+import org.spongepowered.api.util.Color;
+import org.spongepowered.api.util.annotation.NonnullByDefault;
+import org.spongepowered.asm.mixin.Final;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.common.interfaces.block.material.IMixinMapColor;
+import org.spongepowered.common.map.SpongeMapColor;
+
+import java.util.HashMap;
+import java.util.Map;
+
+@NonnullByDefault
+@Mixin(net.minecraft.block.material.MapColor.class)
+public class MixinMapColor implements MapColor.Base, IMixinMapColor {
+
+    @Shadow @Final public int colorValue;
+    @Shadow @Final public int colorIndex;
+
+    private Map<MapShade, MapColor> cachedShadings = new HashMap<>();
+    private String id;
+
+    @Override
+    public Color getColor() {
+        return Color.ofRgb(colorValue);
+    }
+
+    @Override
+    public MapColor shade(MapShade newShading) {
+        if (cachedShadings.containsKey(newShading)) {
+            return cachedShadings.get(newShading);
+        } else {
+            MapColor shaded = new SpongeMapColor(this, newShading);
+            cachedShadings.put(newShading, shaded);
+            return shaded;
+        }
+    }
+
+    @Override
+    public int getContentVersion() {
+        return 1;
+    }
+
+    @Override
+    public DataContainer toContainer() {
+        final DataContainer container = new MemoryDataContainer();
+        container.set(Queries.CONTENT_VERSION, getContentVersion())
+                .set(DataQuery.of("ColorIndex"), this.colorIndex);
```

Comment:
As before, store this as a static final field, preferably in `DataConstants or DataQueries whatever the class it is.